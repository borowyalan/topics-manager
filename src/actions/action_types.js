export const OPEN_GITHUB_LOGIN = 'OPEN_LOGIN_PAGE';
export const AUTH_LOGIN_REQUEST_SUCCESS = 'AUTH_LOGIN_REQUEST_SUCCESS';
export const AUTH_LOGIN_REQUEST_ERROR = 'AUTH_LOGIN_REQUEST_ERROR';
export const LOGOUT = 'LOGOUT';
