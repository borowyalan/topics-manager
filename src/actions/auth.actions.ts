import { AuthService } from '../shared/services/AuthService';
import {
    LOGIN_REQUEST,
    LOGIN_REQUEST_ERROR,
    LOGIN_REQUEST_SUCCESS,
    LOGIN_RESTORE_SESSION_REQUEST,
    LOGOUT_REQUEST,
    LOGOUT_REQUEST_ERROR,
    LOGOUT_REQUEST_SUCCESS,
} from './action_types';

export const requestLogin = () => {
    return dispatch => {
        dispatch(loginPending());
        AuthService.signIn()
            .then(result => {
                dispatch(loginSuccess(result));
            })
            .catch(error => {
                dispatch(loginError(error));
            });
    };
};

export const loginPending = () => {
    return {
        type: LOGIN_REQUEST,
    };
};

export const loginError = error => {
    return {
        type: LOGIN_REQUEST_ERROR,
        payload: error,
    };
};

export const loginSuccess = user => {
    return {
        type: LOGIN_REQUEST_SUCCESS,
        payload: user,
    };
};

export const requestLogout = () => {
    return dispatch => {
        dispatch(logoutPending());
        AuthService.signOut()
            .then(() => {
                dispatch(logoutSuccess());
            })
            .catch(error => {
                dispatch(logoutError(error));
            });
    };
};

export const logoutPending = () => {
    return {
        type: LOGOUT_REQUEST,
    };
};

export const logoutError = error => {
    return {
        type: LOGOUT_REQUEST_ERROR,
        payload: error,
    };
};

export const logoutSuccess = () => {
    return {
        type: LOGOUT_REQUEST_SUCCESS,
    };
};

export const restoringSession = () => {
    return {
        type: LOGIN_RESTORE_SESSION_REQUEST,
    };
};

export const restoreSession = () => {
    return async dispatch => {
        dispatch(restoringSession());
        try {
            const user = await AuthService.getUser();
            const logged = await AuthService.isLogged(user);
            if (logged) {
                dispatch(loginSuccess(user));
            } // else if just 'not logged'
        } catch (err) {
            dispatch(requestLogout()); // force clean up
        }
    };
};
