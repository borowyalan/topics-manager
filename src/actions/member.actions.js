
import TopicService from '../shared/services/TopicService';
import {
    MEMBER_LEAVE_REQUEST,
    MEMBER_LEAVE_REQUEST_ERROR, MEMBER_LEAVE_REQUEST_SUCCESS,
    MEMBER_SUBMIT_REQUEST,
    MEMBER_SUBMIT_REQUEST_ERROR,
    MEMBER_SUBMIT_REQUEST_SUCCESS
} from './action_types';

export const becomeMember = (topic, user) => {
    return function (dispatch) {
        dispatch(becomeMemberInit());
        return TopicService.attend(topic, user).then(() => {
            dispatch(becomeMemberSucceed());
        }).catch(error => {
            dispatch(becomeMemberError(error));
        });
    };
};

export const becomeMemberInit = () => {
    return {
        type: MEMBER_SUBMIT_REQUEST
    };
};

export const becomeMemberSucceed = () => {
    return {
        type: MEMBER_SUBMIT_REQUEST_SUCCESS,
    };
};

export const becomeMemberError = (error) => {
    return {
        type: MEMBER_SUBMIT_REQUEST_ERROR,
        payload: error
    };
};


export const leaveTopic = (topic, user) => {
    return function (dispatch) {
        dispatch(leaveTopicInit());
        return TopicService.leave(topic, user).then(() => {
            dispatch(leaveTopicSucceed());
        }).catch(error => {
            dispatch(leaveTopicError(error));
        });
    };
};

export const leaveTopicInit = () => {
    return {
        type: MEMBER_LEAVE_REQUEST
    };
};

export const leaveTopicSucceed = () => {
    return {
        type: MEMBER_LEAVE_REQUEST_SUCCESS,
    };
};

export const leaveTopicError = (error) => {
    return {
        type: MEMBER_LEAVE_REQUEST_ERROR,
        payload: error
    };
};
