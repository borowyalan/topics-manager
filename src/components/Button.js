import React, { Component } from 'react';
import styled from 'styled-components';

class Button extends Component {

    render() {
        return (
            this.props.type === 'primary' ?
                <PrimaryButton onClick={this.props.onClick}>{this.props.children}</PrimaryButton>
                : <SecondaryButton onClick={this.props.onClick}>{this.props.children}</SecondaryButton>
        );
    }

}

const BasicButton = styled.button`
    border: none;
`;

const PrimaryButton = BasicButton.extend`
  background-color: red;
  color: #fff;
`;

const SecondaryButton = BasicButton.extend`
  backround-color: yellow;
  color: #000;
`;



export default Button;