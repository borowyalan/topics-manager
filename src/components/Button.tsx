import React, { SFC } from 'react';
import styled from 'styled-components';
import Colors from '../styles/colors';

export enum ButtonType {
    Primary = 'primary',
    Secondary = 'secondary',
}

export interface IButtonProps {
    kind?: ButtonType,
}

type Properties = IButtonProps & React.AllHTMLAttributes<HTMLButtonElement>

const Button: SFC<Properties> = (props) => {
    switch (props.kind) {
        case ButtonType.Primary:
            return <PrimaryButton {...props} onClick={props.onClick}>
                {props.children}
            </PrimaryButton>;
        case ButtonType.Secondary:
            return <SecondaryButton {...props} onClick={props.onClick}>
                {props.children}
            </SecondaryButton>;
        default:
            return null;
    }
};

const BasicButton = styled.button`
    border: none;
    padding: 15px;
    font-size: 14px;
    min-width: 180px;
    text-transform: uppercase;
    &:hover {
        box-shadow: 3px 7px 20px 0 rgba(0, 0, 0, 0.2);
        cursor: pointer;
    }
`;

const PrimaryButton = BasicButton.extend`
    background-color: ${Colors.red};
    color: ${Colors.white};
`;

const SecondaryButton = BasicButton.extend`
    background-color: ${Colors.yellow};
    color: ${Colors.black};
`;

export default Button;
