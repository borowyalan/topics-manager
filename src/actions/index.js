import { LOGOUT } from '../actions/action_types';
import { AUTH_LOGIN_REQUEST_SUCCESS } from './action_types';

export const openLoginPage = (history) => {
    return function(dispatch) {
        history.push('/projects');
        setTimeout(() => {
            //pretend we were logging in...
            dispatch(loggedIn());
        }, 1500);
    };
};

export const loggedIn = () => {
    return {
        type: AUTH_LOGIN_REQUEST_SUCCESS
    };
};

export const logout = () => {
    return {
        type: LOGOUT
    };
};
