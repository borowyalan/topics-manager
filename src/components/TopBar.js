import React from 'react';
import styled from 'styled-components';
import Button from './Button';
import { connect } from 'react-redux';
import { openLoginPage, logout } from './../actions';
import Text from './Text';
import Colors from '../styles/Colors';
import PropTypes from 'prop-types';

class TopBar extends React.Component {
    
    onClick = () => {
        const { loggedUser, openLoginPage, logout } = this.props;
        const onPressAction = loggedUser ? logout : openLoginPage;
        onPressAction();
    };
    
    render() {
        const { loggedUser } = this.props;
        const buttonText = loggedUser ? 'Wyloguj' : 'Zaloguj';
        return (
            <TopBarWrapper>
                <Text type="basic" color={ Colors.white } display={ 'inline' }>
                    Zaloguj siÄ™ przez GitHub
                </Text>
                <Button marginLeft='10px' onClick={this.onClick} type="primary">
                    { buttonText }
                </Button>
            </TopBarWrapper>
        );
    }
}

TopBar.propTypes = {
    loggedUser: PropTypes.bool,
    login: PropTypes.func,
    logout: PropTypes.func,
    history: PropTypes.any
};

const mapStateToProps = state => ({
    loggedUser: state.loggedUser
});

const mapDispatchToProps = dispatch => ({
    openLoginPage: () => dispatch(openLoginPage()),
    logout: () => dispatch(logout())
});

const TopBarWrapper = styled.section`
    width: 100%;
    background-color: ${Colors.black};
    color: ${Colors.white};
    padding: 7px;
`;

export default connect(mapStateToProps, mapDispatchToProps)(TopBar);
