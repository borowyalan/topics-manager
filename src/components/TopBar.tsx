import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { requestLogin as loginAction, requestLogout as logoutAction } from '../actions/auth.actions';
import ActivityIndicator from '../shared/components/ActivityIndicator';
import { BackgroundImage } from '../shared/components/BackgroundImage';
import { User } from '../shared/models/user';
import Colors from '../styles/Colors';
import Button, { ButtonType } from './Button';
import Section from './Section';
import Text, { TextType } from './Text';

export interface TopBarProps {
    logged: boolean,
    pending: boolean,
    error?: any,
    user: User,
    requestLogin: () => undefined,
    requestLogout: () => undefined,
}

const ButtonPosition = styled.div`
    margin-left: 15px
`;

class TopBar extends React.Component<TopBarProps> {
    public onClick = () => {
        const { logged, requestLogout, requestLogin } = this.props;
        const onPressAction = logged ? requestLogout : requestLogin;
        onPressAction();
    };

    public render() {
        const { logged, pending, user, error } = this.props;
        const buttonText = logged ? 'Wyloguj' : 'Zaloguj';
        return (
            <Section backgroundColor={Colors.black}>
                <SpaceBetweenContainer>
                    <BackgroundImage url={'http://'}/>
                    <Text kind={TextType.Default}/>
                    <Text kind={TextType.Logo} style={{margin: "0 20px 0 0"}}>
                        Warsaw<strong>JS</strong>
                    </Text>
                    <MenuList>
                        <MenuLink>Meetup</MenuLink>
                        <MenuLink>Worhshop</MenuLink>
                        <MenuLink>Praca</MenuLink>
                        <MenuLink>O nas</MenuLink>
                        <MenuLink>Archiwum</MenuLink>
                        <MenuLink>Formularz</MenuLink>
                        <MenuLink>Kontakt</MenuLink>
                        <MenuLink>Blog</MenuLink>
                    </MenuList>
                    <Text kind={TextType.Basic} color={Colors.white}>
                        {logged && user && user.email}
                    </Text>
                    {pending && <ActivityIndicator/>}
                    {error && <p>Błąd: {JSON.stringify(error)}</p>}

                    <ButtonPosition>
                        <Button
                            onClick={this.onClick}
                            kind={ButtonType.Primary}>
                            {buttonText}
                        </Button>
                    </ButtonPosition>
                </SpaceBetweenContainer>
            </Section>
        );
    }
}

const MenuLink = styled.li`
    color: white;
    &:hover {
        color: ${Colors.red};
        padding: 5px 0;
        cursor: pointer;
        border-bottom: solid 1px ${Colors.white};
    }
`;
const SpaceBetweenContainer = styled.div`
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
`;
const MenuList = styled.ul`
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
    list-style-type: none;
    margin: 0 10px;
`;

const mapStateToProps = state => ({
    logged: state.auth.logged,
    pending: state.auth.pending,
    user: state.auth.user,
    error: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
    requestLogin: () => dispatch(loginAction()),
    requestLogout: () => dispatch(logoutAction()),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(TopBar);
