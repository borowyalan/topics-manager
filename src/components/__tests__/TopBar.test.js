import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';
import TopBar from '../TopBar';
import Button from '../Button';
import { LOGIN, LOGOUT } from '../../actions/action_types';

configure({ adapter: new Adapter() });
const mockStore = configureMockStore();

describe('<TopBar>', () => {
    const initialState = {
        auth: {
            loggedUser: false
        }
    };
    let wrapper;
    let store = mockStore(initialState);
    
    beforeEach(() => {
        wrapper = shallow(<TopBar store={store}/>).dive();
    });
    
    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    describe('given that user is logged', () => {
        let button;
        beforeEach(() => {
            store = mockStore({
                auth: {
                    loggedUser: true
                }
            });
            wrapper = shallow(<TopBar store={store}/>).dive();
            button = wrapper.find(Button);
        });
        
        it('should have a proper text on button', () => {
            expect(button.children().text()).toBe('Wyloguj');
        });
        
        it('should dispatch logout action on click', () => {
            button.simulate('click');
            const actions = store.getActions();
            expect(actions.length).toBe(1);
            expect(actions[0].type).toBe(LOGOUT);
        });
        
    });
    
    describe('given that user is not logged in', () => {
        let button;
        beforeEach(() => {
            store = mockStore({
                auth: {
                    loggedUser: false
                }
            });
            wrapper = shallow(<TopBar store={store}/>).dive();
            button = wrapper.find(Button);
        });
        
        it('should have a proper text on button', () => {
            expect(button.children().text()).toBe('Zaloguj');
        });
        
        it('should dispatch login action on click', () => {
            button.simulate('click');
            const actions = store.getActions();
            expect(actions.length).toBe(1);
            expect(actions[0].type).toBe(LOGIN);
        });
    });
    
});
