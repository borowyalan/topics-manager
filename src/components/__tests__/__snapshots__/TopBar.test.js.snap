// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopBar> should render 1`] = `
<styled.section>
  <Text
    color="#fff"
    display="inline"
    type="basic"
  >
    Zaloguj siÄ™ przez GitHub
  </Text>
  <Button
    marginLeft="10px"
    onClick={[Function]}
    type="primary"
  >
    Zaloguj
  </Button>
</styled.section>
`;
